poly/schnorr.go

```diff
@@ -0,0 +1,270 @@
+package poly
+
+import (
+	"errors"
+	"fmt"
+	"github.com/dedis/crypto/abstract"
+)
+
+// This file describe the Distributed Threshold Schnorr Signature
+
+// Schnorr will holds the data necessary to complete a distributed schnorr signature
+// And will implement methods to do so
+// Basically you can setup a schnorr struct with a LongTerm shared secret
+// When you want to sign something, you
+//  - Start a new round specifying the random shared secret chosen and the message to sign
+//  - Generate the partial signature of the current node
+//  - Collect every others partial signature
+//  - Generate signature
+//  - ... Do what ever you want to do with
+//  - You can now start again a new round with the same schnorr struct
+// When you want to verify a given signature :
+//  - You can just call from the schnorr struct : schnorr.VerifySignature(SchnorrSig, msg)
+// CAREFUL: your schnorr signature is a LONG TERM signature, you must keep the same throughout the differents rounds
+// otherwise you won't be able to verify any signatures. To be more precise, you just have to keep the same LongTerm sharedSecret
+// and PolyInfo. If you know these are the same throughout differents rounds, you can create many schnorr structs. This is
+// definitly NOT the way it is intented to be used, so use it at your own risks.
+type Schnorr struct {
+
+	// The info describing which kind of polynomials we using, on which groups etc
+	info PolyInfo
+
+	// The long-term shared secret evaluated by receivers
+	Longterm *SharedSecret
+
+	////////////////////////////////////////////////////
+	// FOR A GIVEN ROUND, we have the following members :
+
+	// hash is the hash of the message
+	hash *abstract.Secret
+
+	// The short term shared secret to use for this signature ONLY /!\
+	// i.e. the random secret in the regular schnorr signature
+	random *SharedSecret
+
+	// The partials signature of each other peers (i.e. receiver)
+	partials []*PartialSchnorrSig
+
+	/////////////////////////////////////////////////////
+}
+
+// Partial Schnorr Sig represents the partial signatures that each peers must generate in order to
+// generate the "global" signature. This struct must be sent across each peers for each peers
+type PartialSchnorrSig struct {
+	// The index of this partial signature regarding the global one
+	// same as the "receiver" index in the joint.go code
+	Index int
+
+	// The partial signature itself
+	Part *abstract.Secret
+}
+
+// SchnorrSig represents the final signature of a distribtued threshold schnorr signature
+// which can be verified against a message
+// This struct is not intended to be constructed manually but can be:
+//  - produce by the Schnorr struct
+//  - verified against a Schnorr struct
+type SchnorrSig struct {
```

Comment:
If this is only supposed to be treated as an opaque blob by clients of this package, shouldn't it be private?  Are the fields public because it's used for marshaling/unmarshaling?  If so, it might be better for the clients just to see []byte-slices, and have this be a private struct used only internally for marshaling.  (Note that private structs can have public fields and can be used for mashaling/unmarshaling.)